server/api/animelist.ts

```diff
@@ -0,0 +1,174 @@
+import axios from 'axios';
+import xml2js from 'xml2js';
+import fs, { promises as fsp } from 'fs';
+import path from 'path';
+import logger from '../logger';
+
+const UPDATE_INTERVAL = 24 * 3600; // how often to download new mapping
+// originally at https://raw.githubusercontent.com/ScudLee/anime-lists/master/anime-list.xml
+const MAPPING_URL =
+  'https://raw.githubusercontent.com/Anime-Lists/anime-lists/master/anime-list.xml';
+const LOCAL_PATH = path.join(__dirname, '../../config/anime-list.xml');
+
+const mappingRegexp = new RegExp(/;[0-9]+-([0-9]+);/);
+
+// Anime-List xml files are community maintained mappings that Hama agent uses to map AniDB IDs to tvdb/tmdb IDs
+// https://github.com/Anime-Lists/anime-lists/
+
+export interface AnidbItem {
+  tvdbId?: number;
+  tmdbId?: number;
+  imdbId?: string;
+}
+
+class AnimeListMapping {
+  private syncing = false;
+
+  private mapping: { [anidbId: number]: AnidbItem } = {};
+
+  // each episode in season 0 from TVDB can map to movie (AniDB id)
+  private specials: { [tvdbId: number]: { [episode: number]: number } } = {};
+
+  public isLoaded = () => Object.keys(this.mapping).length !== 0;
+
+  private loadFromFile = async () => {
+    logger.info('Loading Anime-List mapping file');
+    try {
+      const file = await fsp.readFile(LOCAL_PATH);
+      const xml = await xml2js.parseStringPromise(file);
+
+      this.mapping = {};
+      this.specials = {};
+      for (const anime of xml['anime-list']['anime']) {
+        const a = anime['$'];
+
+        let tvdbId = a['tvdbid'];
+        if (!isNaN(Number(tvdbId))) {
+          tvdbId = Number(tvdbId);
+        } else {
+          tvdbId = undefined;
+        }
+
+        let imdbId = a['imdbid'];
+        if (imdbId && !imdbId.startsWith('tt')) {
+          imdbId = undefined;
+        }
+
+        const tmdbId = a['tmdbid'];
+        const anidbId = Number(a['anidbid']);
+        this.mapping[anidbId] = {
+          tvdbId: tvdbId,
+          tmdbId: tmdbId,
+          imdbId: imdbId,
+        };
+
+        if (tvdbId) {
+          const mappingLists = anime['mapping-list'];
+          if (mappingLists) {
+            for (const mappingList of mappingLists) {
+              for (const mapping of mappingList['mapping']) {
+                const text = mapping['_'];
+                if (text && mapping['$']['tvdbseason'] == '0') {
```

Comment:
Strict comparison here

```suggestion
                if (text && mapping['$']['tvdbseason'] === '0') {
```